### PR TITLE
Flatten dims with scalars in MultiIndex

### LINKED ISSUE #3367: Flatten dims in operations with scalars
Continuation of #3365

Given, e.g.,
```python
sc.ones(sizes={'x': 2, 'y': 3}) + sc.scalar(1)
```
should we flatten the dims to effectively compute the following?
```python
sc.fold(sc.ones(sizes={'x': 2, 'y': 3}).flatten(to='aux') + sc.scalar(1), sizes={'x': 2, 'y': 3})
```
This should be possible in `MultiIndex`. And similarly for length 0 dims in any operand.

It would extend the usability of arrays with more than 6 dims.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #3367

### CODE REVIEW

File: lib/variable/test/variable_test.cpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ TEST(Variable, many_dims_works_or_fails_gracefully) {
   EXPECT_EQ(var + var, makeVariable<double>(dims, Values{2}));
   // TODO In principle we should be able to support all of the below with
   // flattening, but the current implementation dos not handle this.
-  ASSERT_THROW(var += 1.0 * units::one, std::runtime_error);
+  EXPECT_EQ(var + 1.0 * units::one, makeVariable<double>(dims, Values{2}));
```
Comment (SimonHeybrock) [inline lib/variable/test/variable_test.cpp:42]: Move above the TODO?

File: lib/dataset/test/binned_arithmetic_test.cpp [hunk 2]
```diff
@@ -111,13 +111,15 @@ TEST_F(BinnedArithmeticTest, op_on_slice) {
   EXPECT_EQ(slice * one, copy(slice));
 }
 
-TEST_F(BinnedArithmeticTest, too_many_dims) {
+TEST_F(BinnedArithmeticTest, many_dims) {
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/dataset/test/binned_arithmetic_test.cpp:114]: How about changing one or a couple dims to length 2, and then operate on a slice? This should prevent flattening.

**Comment by jl-wynen** [comment]
> How about changing one or a couple dims to length 2, and then operate on a slice? This should prevent flattening.

I tried that and it didn't work. I think because it flattens the length-1 dims. Maybe if all dims where sliced. But I don't know a good way of making that data that doesn't involve dozens of lines of array creation code.
```cpp
  const auto indices6d = makeVariable<scipp::index_pair>(
      Dims{Dim{"a"}, Dim{"b"}, Dim{"c"}, Dim{"d"}, Dim{"e"}, Dim{"f"}},
      Shape{1, 1, 2, 1, 2, 1}, Values{std::pair{0, 1}, std::pair{1, 2}, std::pair{2, 3}, std::pair{3, 4}});
  const auto buffer = makeVariable<int>(Dims{Dim::Event}, Shape{4}, Values{1, 2, 3, 4});
  const auto binned = make_bins(indices6d, Dim::Event, buffer).slice({Dim{"c"}, 0, 1}).slice({Dim{"e"}, 0, 1});
  const auto rhs = makeVariable<int>(Dims{}, Values{3});
  EXPECT_THROW_DISCARD(binned + rhs, std::runtime_error);
```

**Comment by SimonHeybrock** [comment]
> > How about changing one or a couple dims to length 2, and then operate on a slice? This should prevent flattening.
> 
> I tried that and it didn't work. I think because it flattens the length-1 dims. Maybe if all dims where sliced. But I don't know a good way of making that data that doesn't involve dozens of lines of array creation code.
> 
> ```c++
>   const auto indices6d = makeVariable<scipp::index_pair>(
>       Dims{Dim{"a"}, Dim{"b"}, Dim{"c"}, Dim{"d"}, Dim{"e"}, Dim{"f"}},
>       Shape{1, 1, 2, 1, 2, 1}, Values{std::pair{0, 1}, std::pair{1, 2}, std::pair{2, 3}, std::pair{3, 4}});
>   const auto buffer = makeVariable<int>(Dims{Dim::Event}, Shape{4}, Values{1, 2, 3, 4});
>   const auto binned = make_bins(indices6d, Dim::Event, buffer).slice({Dim{"c"}, 0, 1}).slice({Dim{"e"}, 0, 1});
>   const auto rhs = makeVariable<int>(Dims{}, Values{3});
>   EXPECT_THROW_DISCARD(binned + rhs, std::runtime_error);
> ```

How about adding the binned data to a copy of its transpose?

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**